Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1391 @@
+    case .clientOpenServerOpen(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientOpenServerOpen(state)
+      return request.map { .receiveMessage($0) } ?? .awaitMoreMessages
+    case .clientOpenServerClosed(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientOpenServerClosed(state)
+      return request.map { .receiveMessage($0) } ?? .awaitMoreMessages
+    case .clientClosedServerOpen(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientClosedServerOpen(state)
+      return request.map { .receiveMessage($0) } ?? .noMoreMessages
+    case .clientClosedServerClosed(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientClosedServerClosed(state)
+      return request.map { .receiveMessage($0) } ?? .noMoreMessages
+    case .clientClosedServerIdle:
+      return .noMoreMessages
+    case .clientIdleServerIdle:
+      return .awaitMoreMessages
+    }
+  }
+}
+
+extension MethodDescriptor {
+  init?(fullyQualifiedMethod: String) {
+    let split = fullyQualifiedMethod.split(separator: "/")
+    guard split.count == 2 else {
+      return nil
+    }
+    self.init(service: String(split[0]), method: String(split[1]))
+  }
+}
+
+internal enum GRPCHTTP2Keys: String {
+  case path = ":path"
+  case contentType = "content-type"
+  case encoding = "grpc-encoding"
+  case acceptEncoding = "grpc-accept-encoding"
+  case scheme = ":scheme"
+  case method = ":method"
+  case te = "te"
+  case status = ":status"
+  case grpcStatus = "grpc-status"
+  case grpcStatusMessage = "grpc-status-message"
+}
+
+extension HPACKHeaders {
+  internal func firstString(forKey key: GRPCHTTP2Keys) -> String? {
+    self.values(forHeader: key.rawValue, canonicalForm: true).first(where: { _ in true }).map {
+      String($0)
+    }
+  }
+
+  internal mutating func add(_ value: String, forKey key: GRPCHTTP2Keys) {
+    self.add(name: key.rawValue, value: value)
+  }
+}
+
+extension Zlib.Method {
+  init?(encoding: CompressionAlgorithm) {
+    switch encoding {
+    case .identity:
+      return nil
+    case .deflate:
+      self = .deflate
+    case .gzip:
+      self = .gzip
+    default:
+      return nil
+    }
+  }
+}
+
+extension Metadata {
+  init(headers: HPACKHeaders) {
+    var metadata = Metadata()
+    for header in headers {
+      if header.name.hasSuffix("-bin") {
+        do {
+          let decodedBinary = try header.value.base64Decoded()
+          metadata.addBinary(decodedBinary, forKey: header.name)
+        } catch {
+          metadata.addString(header.value, forKey: header.name)
+        }
+      } else {
+        if header.name == GRPCHTTP2Keys.grpcStatusMessage.rawValue {
+          let decodedStatusMessage = GRPCStatusMessageMarshaller.unmarshall(header.value)
+          metadata.addString(decodedStatusMessage, forKey: header.name)
```

Comment:
Hmm, didn't we have code elsewhere to remove values for this and the status code? If so there's no point special casing it here.